### ISSUE TITLE
Is it possible to upgrade two composer dependencies in the same command?

### ISSUE THREAD

**Issue by tobybellwood** [renovatebot/config-help#883]
WhiteSource Renovate App

### Which platform are you using?

GitHub.com

### Have you checked the logs? Don't forget to include them if relevant

https://app.renovatebot.com/dashboard#github/tobybellwood/drupal-example-simple/236661111

### What would you like to do?

There is a particular condition seen in Drupal that shares a composer package (drupal/core) between require and require-dev.  If the two repos aren't composer updated together, they will  not reconcile and we get the Artifact update problem

As outlined in the logs:
```
  Problem 1
    - drupal/core-recommended 9.0.1 requires drupal/core 9.0.1 -> satisfiable by drupal/core[9.0.1].
    - drupal/core-recommended 9.0.1 requires drupal/core 9.0.1 -> satisfiable by drupal/core[9.0.1].
    - Can only install one of: drupal/core[9.0.6, 9.0.1].
    - Can only install one of: drupal/core[9.0.6, 9.0.1].
    - drupal/core-dev-pinned 9.0.6 requires drupal/core 9.0.6 -> satisfiable by drupal/core[9.0.6].
    - Installation request for drupal/core-dev-pinned 9.0.6 -> satisfiable by drupal/core-dev-pinned[9.0.6].
    - Installation request for drupal/core-recommended 9.0.1 -> satisfiable by drupal/core-recommended[9.0.1].
```

Usually, the two are upgraded together by:
`composer update drupal/core-recommended drupal/core-dev-pinned --with-dependencies`

Is there a way to get them to update together - i thought groups might do it in Renovate, but I think that only groups PRs together, not processes them together?

I will experiment a little more

**Comment by rarkins** [comment]
I can see it would be a problem if we run "composer update" commands one by one instead of at the end. But do we quit and give up after the first error, or merely log it?

A public reproduction repo would be necessary for us to file a bug or feature request if you can't work out a way to resolve it.

**Comment by rarkins** [comment]
BTW I thought that our existing approach and grouping would work for you already though, so reproducing is important.

**Comment by tobybellwood** [comment]
I'm running one now - I pinned both to the same, and put them into a group over at
https://github.com/tobybellwood/drupal-example-simple/blob/9.x/renovate.json

**Comment by tobybellwood** [comment]
Righto - don't mind me.

Once I pinned both to the same release and grouped them they update together perfectly - https://app.renovatebot.com/dashboard#github/tobybellwood/drupal-example-simple/236685608

The caveat is that this won't work for an initial semver>pin PR, only once they've all been pinned.

**Comment by rarkins** [comment]
Why wouldn't it work? The pinning should be done to the same version as in the lock file so would product the exact same result.

**Comment by tobybellwood** [comment]
Yeah - I'm not sure either

From the renovate logs, it looks like it didn't read in a lockedVersion for drupal/core-dev-pinned
```
          {
            "depType": "require-dev",
            "depName": "drupal/core-dev-pinned",
            "currentValue": "^9.0",
            "datasource": "packagist",
            "depIndex": 8,
            "updates": [
              {
                "updateType": "pin",
                "isPin": true,
                "newValue": "9.0.6",
                "newMajor": 9
              }
            ],
            "warnings": [],
            "sourceUrl": "https://github.com/drupal/core-dev-pinned"
          }
```
Even though it was pinned in the composer.lock at the time - https://github.com/tobybellwood/drupal-example-simple/blob/2c957f7999e4846ac2966485cf756e49ae8d0e13/composer.lock#L6472
```
            "name": "drupal/core-dev-pinned",
            "version": "9.0.1",
```
Which would explain why it tried to pin forwards to 9.0.6, but the others read the lockedVersion at 9.0.6.

It's a bit chicken-and-egg - the renovate group doesn't exist before the pin dependencies PR is created, but knowing why it didn't read the lockedVersion in (require-dev maybe?) would seem to be the root cause

**Comment by rarkins** [comment]
Yes we should make sure to read lockedVersion. Is that something you can reproduce?

**Comment by tobybellwood** [comment]
I'll see if I can create another one.  I've reread the renovate code and can't see why it didn't pick it up from composer.lock

**Comment by tobybellwood** [comment]
i've just added a whole slew of new baseBranches to my existing renovate.json (with the groups already in place) - but the Pin dependency PRs don't pick up the lockedVersion for the require-dev dependency (PRs at https://github.com/tobybellwood/drupal-example-simple/pulls (7,11,12,13,14) and job run at https://app.renovatebot.com/dashboard#github/tobybellwood/drupal-example-simple/237168747

**Comment by tobybellwood** [comment]
I've also added another require-dev dependency into a branch and re-run an initial Pin dependency, and it too missed the LockVersion, and tried to pin ahead (composer.lock at 1.7.0)
```
          {
            "depType": "require-dev",
            "depName": "pyrech/composer-changelogs",
            "currentValue": "^1.7",
            "datasource": "packagist",
            "depIndex": 12,
            "updates": [
              {
                "updateType": "pin",
                "isPin": true,
                "newValue": "1.7.1",
                "newMajor": 1
              }
            ],
            "warnings": [],
            "sourceUrl": "https://github.com/pyrech/composer-changelogs"
          }
        ],
```
https://github.com/tobybellwood/drupal-example-simple/pull/15

**Comment by rarkins** [comment]
It's a little big/complex for debugging but hopefully not too bad. Assuming you are now satisfied with the reproduction, and can leave it as-is so others in future can fork it as-is, can you clarify:

- Which PR to look at? e.g. 15?
- Which dependencies are pinned correctly, and which are pinned incorrectly? For the incorrect ones, what is the version you'd expect instead?

**Comment by tobybellwood** [comment]
Totally - I can maybe make a simpler one now that my hypothesis (require-dev not reading LockedVersions) seems to be the problem.

Basically in PR15:
* renovate_test branch has pyrech/composer-changelogs at v1.7.0 in composer.lock 
https://github.com/tobybellwood/drupal-example-simple/blob/renovate_test/composer.lock#L8247
* renovate creates a Pin dependencies PR (https://github.com/tobybellwood/drupal-example-simple/pull/15) that wants to pin it to v1.7.1 instead (latest matching release)

pyrech/composer-changelogs | require-dev | pin | ^1.7 -> 1.7.1
-- | -- | -- | --

This is caused by Renovate seemingly not picking up the LockedVersion (as v1.7.0) from the composer.lock - possible a require-dev issue?
There is another require-dev in the same project/branch (drupal/core-dev-pinned) that does the same (composer.lock pinned at 9.0.1, Renovate wants to pin to 9.0.6 - latest matching release)

**Comment by tobybellwood** [comment]
ok - much simpler example that proves this - https://github.com/tobybellwood/composer_demo/pull/1
